readme.md
# Circos [![Circle CI](https://circleci.com/gh/nicgirault/circosJS.svg?style=shield)](https://circleci.com/gh/nicgirault/circosJS) [![Coverage Status](https://coveralls.io/repos/github/nicgirault/circosJS/badge.svg?branch=master)](https://coveralls.io/github/nicgirault/circosJS?branch=master)

Demo: [nicgirault.github.io/circosJS/demo](https://nicgirault.github.io/circosJS/demo)

## table of contents

- [Introduction](#introduction)
- [Installation](#installation)
- [Layout](#layout)
- [Tracks](#tracks)
  - [Chords](#chords)
  - [Heatmap](#heatmap)
  - [Highlight](#highlight)
  - [Histogram](#histogram)
  - [Line](#line)
  - [Scatter](#scatter)
  - [Stack](#stack)
  - [Text](#text)
- [Configuration Attributes](#configurationAttributes)
  - [innerRadius](#innerRadiusOuterRadius)
  - [outerRadius](#innerRadiusOuterRadius)
  - [min](#minmax)
  - [max](#minmax)
  - [color](#color)
  - [axes](#axes)
  - [showAxesTooltip](#showAxesTooltip)

## Introduction

Circos is a javascript library to easily build interactive graphs in a circular layout. It's based on [d3.js](https://d3js.org/). It aims to be a javascript version of the [Circos](http://circos.ca) software.

You should consider using Circos to show:

- relationships between entities
- periodical data

<p align="center">
  <img src="doc/temperatures.png" width="60%" alt="temperatures">
  <br/>
  <i>Average temperatures in Paris from 2007 (inner) to 2014 (outer). The circular layout highlights seasonal effect.</i>
</p>

## Installation

If you don't know what is yarn or npm you can skip this step and get started with [this canvas](doc/canvas.html). Otherwise:

```
yarn add circos
```

## Layout

To instantiate a new circos:

```javascript
var myCircos = new Circos({
    container: '#chart',
    width: 500,
    height: 500,
});
```

A circos graph is based on a circular axis **layout**. Data tracks appear inside and/or outside the circular layout.

In order to place data on the circos graph, you must first specify the layout.

```javascript
myCircos.layout(data, configuration);
```

The first argument of the `layout` function is a configuration object that control the format of the layout.

Here are the default parameters for a layout:

```javascript
var configuration = {
  innerRadius: 250,
  outerRadius: 300,
  cornerRadius: 10,
  gap: 0.04, // in radian
  labels: {
    display: true,
    position: 'center',
    size: '14px',
    color: '#000000',
    radialOffset: 20,
  },
  ticks: {
    display: true,
    color: 'grey',
    spacing: 10000000,
    labels: true,
    labelSpacing: 10,
    labelSuffix: 'Mb',
    labelDenominator: 1000000,
    labelDisplay0: true,
    labelSize: '10px',
    labelColor: '#000000',
    labelFont: 'default',
    majorSpacing: 5,
    size: {
      minor: 2,
      major: 5,
    }
  },
  clickCallback: null
}
```

The second argument of the `layout` function is an array of data that describe the layout regions. Each layout region must have an id and a length. You can also specify a color and a label.

```javascript
var data = [
  { len: 31, color: "#8dd3c7", label: "January", id: "january" },
  { len: 28, color: "#ffffb3", label: "February", id: "february" },
  { len: 31, color: "#bebada", label: "March", id: "march" },
  { len: 30, color: "#fb8072", label: "April", id: "april" },
  { len: 31, color: "#80b1d3", label: "May", id: "may" },
  { len: 30, color: "#fdb462", label: "June", id: "june" },
  { len: 31, color: "#b3de69", label: "July", id: "july" },
  { len: 31, color: "#fccde5", label: "August", id: "august" },
  { len: 30, color: "#d9d9d9", label: "September", id: "september" },
  { len: 31, color: "#bc80bd", label: "October", id: "october" },
  { len: 30, color: "#ccebc5", label: "November", id: "november" },
  { len: 31, color: "#ffed6f", label: "December", id: "december" }
]
```

The `id` parameter will be used to place data points on the layout.

To visualize the result:

```javascript
myCircos.render();
```


## Tracks

A track is a series of data points.

To add a track to your graph you should write something like this:

```javascript
myCircos.heatmap(
    'my-heatmap',
    data,
    {
        // your heatmap configuration (optional)
    },
);
```

This pattern is similar to all track types:

```javascript
myCircos.trackType('track-id', data, configuration);
```

**Note**: The track name is used as a HTML class name so here are the format limitations.

* Must be unique.
* Should be slug style for simplicity, consistency and compatibility. Example: `my-heatmap`
* Lowercase, a-z, can contain digits, 0-9, can contain dash or dot but not start/end with them.
* Consecutive dashes or dots not allowed.
* 50 characters or less.


### Chords

Chords tracks connect layout regions.

<p align="center">
  <img src="doc/chords.png" width="60%" alt="chords">
  <br/>
  <i>Some gene fusions in human karyotype (<a href="demo/chords.js">source</a>)</i>
</p>

Data should looks like this:

```javascript
var data = [
    // sourceId, sourceStart, sourceEnd, targetId, targetStart, targetEnd
    ['january', 1, 12, 'april', 18, 20],
    ['february', 20, 28, 'december', 1, 13],
];
```

Optionally each datum can define a seventh element which can be used to be interpreted as a `value` to draw colored ribbons with palettes or a color function.

The default configuration is:

```javascript
{
  color: '#fd6a62',
  opacity: 0.7,
  zIndex: 1,
  tooltipContent: null,
  min: null,
  max: null,
  logScale: false,
  logScaleBase: Math.E,
}
```

### Heatmap

<p align="center">
  <img src="doc/heatmap.png" width="60%" alt="heatmap">
  <br/>
  <i>Electrical comsumption in France in 2014</i>
</p>

To add a heatmap to your circos instance:

```javascript
myCircos.heatmap('electrical-consumption', data, configuration);
```

Configuration:

```javascript
{
  innerRadius: null,
  outerRadius: null,
  min: null,
  max: null,
  color: 'YlGnBu',
  logScale: false,
  tooltipContent: null,
}
```

Data format:

```javascript
var data = [
    // each datum should be
    // layout_block_id, start, end, value
    ['january', 0, 1, 1368001],
    ['january', 1, 2, 1458583],
    ['january', 2, 3, 1481633],
    ['january', 3, 4, 1408424]
    ...
    ['february', 0, 1, 1577419],
    ['february', 1, 2, 1509311],
    ['february', 2, 3, 1688266],
    ...
]
```

### Highlight

<p align="center">
  <img src="doc/highlight.png" width="60%" alt="highlight">
  <br/>
  <i>Human karyotype with cytobands highlighted (<a href="demo/highlight.js">source</a>)</i>
</p>

To add a highlight to your circos instance:

```javascript
myCircos.highlight('cytobands', data, configuration);
```

The minimal datum should have `block_id`, `start` and `end` attributes.

Configuration:

```javascript
{
  innerRadius: null,
  outerRadius: null,
  min: null,
  max: null,
  color: 'd3d3d3',
  strokeColor: null,
  strokeWidth: 0,
  opacity: 1,
  logScale: false,
  tooltipContent: null,
}
```

### Histogram

<p align="center">
  <img src="doc/histogram.png" width="60%" alt="histogram">
  <br/>
  <i>Genetic abnormalities in human stem cells (<a href="demo/histogram.js">source</a>)</i>
</p>

Data should looks like this:

```javascript
var data = [
    {
      block_id: 'january',
      start: 1,
      end: 10,
      value: 5
    }
];
```

The available configuration fields are:
- [innerRadius](#innerRadiusOuterRadius)
- [outerRadius](#innerRadiusOuterRadius)
- [color](#color)
- [opacity](#opacity)
- [zIndex](#zIndex)
- [tooltipContent](#tooltipContent)
- [min](#minmax)
- [max](#minmax)
- [logScale](#logScale)
- [logScaleBase](#logScaleBase)
- [axes](#axes)

### Line

<p align="center">
  <img src="doc/line.png" width="60%" alt="line">
  <br/>
  <i>Some single nucleotide polymorphism on chromosomes 1, 2 and 3 (<a href="demo/line.js">source</a>)</i>
</p>

```javascript
myCircos.line('line1', data, configuration);
```

The minimal datum should have `block_id`, `position` and `value` attributes.

The available configuration fields are:
- [innerRadius](#innerRadiusOuterRadius)
- [outerRadius](#innerRadiusOuterRadius)
- [color](#color)
- [strokeColor](#strokeColor)
- [strokeWidth](#strokeWidth)
- [direction](#direction)
- [fill](#fill)
- [maxGap](#maxGap)
- [opacity](#opacity)
- [zIndex](#zIndex)
- [min](#minmax)
- [max](#minmax)
- [logScale](#logScale)
- [logScaleBase](#logScaleBase)
- [axes](#axes)
- [backgrounds](#backgrounds)

**Note**: The tooltip option is not available for line track. To display a tooltip, you should superimpose an invisble `scatter` track (`fill: false` and `strokeWidth: 0`) and set a tooltip for this track.

### Scatter

<p align="center">
  <img src="doc/scatter.png" width="60%" alt="scatter">
  <br/>
  <i><a href="demo/line.js">source</a></i>
</p>

```javascript
myCircos.scatter('scatter1', data, configuration);
```

The minimal datum should have `block_id`, `position` and `value` attributes.

The available configuration fields are:
- [innerRadius](#innerRadiusOuterRadius)
- [outerRadius](#innerRadiusOuterRadius)
- [color](#color)
- [strokeColor](#strokeColor)
- [strokeWidth](#strokeWidth)
- [direction](#direction)
- [fill](#fill)
- [size](#size)
- [shape](#shape)
- [opacity](#opacity)
- [zIndex](#zIndex)
- [min](#minmax)
- [max](#minmax)
- [logScale](#logScale)
- [logScaleBase](#logScaleBase)
- [axes](#axes)
- [backgrounds](#backgrounds)

### Stack

<p align="center">
  <img src="doc/stack.png" width="60%" alt="stack">
  <br/>
  <i><a href="demo/stack.js">source</a></i>
</p>

```javascript
myCircos.stack('stack', data, configuration);
```

The minimal datum should have `block_id`, `start` and `end` attributes.

Configuration:

```javascript
{
  innerRadius: null,
  outerRadius: null,
  min: null,
  max: null,
  color: '#fd6a62',
  strokeColor: '#d3d3d3',
  strokeWidth: 1,
  direction: 'out',
  thickness: 10,
  radialMargin: 2,
  margin: 2,
  opacity: 1,
  logScale: false,
  tooltipContent: null,
}
```

### Text

<p align="center">
  <img src="doc/text.png" width="60%" alt="text">
  <br/>
  <i><a href="demo/text.js">source</a></i>
</p>

```javascript
myCircos.text('text', data, configuration);
```

The minimal datum should have `block_id`, `position` and `value` attributes.

Configuration:

```javascript
{
  innerRadius: null,
  outerRadius: null,
  style: {
    'font-size': 12,
    color: 'black',
  },
  opacity: 1,
}
```

## Configuration Attributes

### backgrounds

You can add a list of backgrounds:

```javascript
{
  backgrounds: [
    {
      start: 0.006,
      color: '#4caf50',
      opacity: 0.1
    },
    {
      start: 0.002,
      end: 0.006,
      color: '#d3d3d3',
      opacity: 0.1
    },
    {
      end: 0.002,
      color: '#f44336',
      opacity: 0.1
    }
  ]
}
```

The `start` and `end` fields are interpreted as values on the same scale than the track values.
- If `start` is not specified, default is the `min` value of the track.
- If `end` is not specified, default is the `max` value of the track.


CHANGELOG.md
# Changelog

### 2.0.3 - 2017-04-05

- Fix tooltip position when we can scroll on the page

### 2.0.2 - 2017-03-16

- Add CTRL+C shortcut to copy tooltip content.
- Add tooltip on axes


demo/demo.js
var layout_data = [
  { "len": 31, "color": "#8dd3c7", "label": "January", "id": "january" },
  { "len": 28, "color": "#ffffb3", "label": "February", "id": "february" },
  { "len": 31, "color": "#bebada", "label": "March", "id": "march" },
  { "len": 30, "color": "#fb8072", "label": "April", "id": "april" },
  { "len": 31, "color": "#80b1d3", "label": "May", "id": "may" },
  { "len": 30, "color": "#fdb462", "label": "June", "id": "june" },
  { "len": 31, "color": "#b3de69", "label": "July", "id": "july" },
  { "len": 31, "color": "#fccde5", "label": "August", "id": "august" },
  { "len": 30, "color": "#d9d9d9", "label": "September", "id": "september" },
  { "len": 31, "color": "#bc80bd", "label": "October", "id": "october" },
  { "len": 30, "color": "#ccebc5", "label": "November", "id": "november" },
  { "len": 31, "color": "#ffed6f", "label": "December", "id": "december" }
];
var heatmap = [
  {
    parent: "chr1",
    data: [
      {start:1, end:10000000, value:0},
      {start:10000001, end:20000000, value:1},
      {start:20000001, end:30000000, value:2},
      {start:30000001, end:40000000, value:3},
      {start:40000001, end:50000000, value:4},
      {start:50000001, end:60000000, value:5},
      {start:60000001, end:70000000, value:6},
      {start:70000001, end:80000000, value:7},
      {start:80000001, end:90000000, value:8},
      {start:90000001, end:100000000, value:9},
      {start:100000001, end:110000000, value:9},
      {start:110000001, end:120000000, value:8},
      {start:120000001, end:130000000, value:7},
      {start:130000001, end:140000000, value:6},
      {start:140000001, end:150000000, value:5},
      {start:150000001, end:160000000, value:4},
      {start:160000001, end:170000000, value:3},
      {start:170000001, end:180000000, value:2},
      {start:180000001, end:190000000, value:1},
      {start:190000001, end:200000000, value:0},
      {start:200000001, end:210000000, value:0},
      {start:210000001, end:220000000, value:1},
      {start:220000001, end:230000000, value:2},
      {start:230000001, end:240000000, value:3},
      {start:240000001, end:249250621, value:3}
    ]
  },
  {
    parent: "chr2",
    data: [
      {start:1, end:10000000, value:4},
      {start:10000001, end:20000000, value:5},
      {start:20000001, end:30000000, value:6},
      {start:30000001, end:40000000, value:7},
      {start:40000001, end:50000000, value:8},
      {start:50000001, end:60000000, value:9},
      {start:60000001, end:70000000, value:9},
      {start:70000001, end:80000000, value:8},
      {start:80000001, end:90000000, value:7},
      {start:90000001, end:100000000, value:6},
      {start:100000001, end:110000000, value:5},
      {start:110000001, end:120000000, value:4},
      {start:120000001, end:130000000, value:3},
      {start:130000001, end:140000000, value:2},
      {start:140000001, end:150000000, value:1},
      {start:150000001, end:160000000, value:0},
      {start:160000001, end:170000000, value:0},
      {start:170000001, end:180000000, value:1},
      {start:180000001, end:190000000, value:2},
      {start:190000001, end:200000000, value:3},
      {start:200000001, end:210000000, value:4},
      {start:210000001, end:220000000, value:5},
      {start:220000001, end:230000000, value:6},
      {start:230000001, end:240000000, value:7},
      {start:240000001, end:243199373, value:7}
    ]
  },
  {
    parent: "chr3",
    data: [
      {start:1, end:10000000, value:8},
      {start:10000001, end:20000000, value:9},
      {start:20000001, end:30000000, value:9},
      {start:30000001, end:40000000, value:8},
      {start:40000001, end:50000000, value:7},
      {start:50000001, end:60000000, value:6},
      {start:60000001, end:70000000, value:5},
      {start:70000001, end:80000000, value:4},
      {start:80000001, end:90000000, value:3},
      {start:90000001, end:100000000, value:2},
      {start:100000001, end:110000000, value:1},
      {start:110000001, end:120000000, value:0},
      {start:120000001, end:130000000, value:0},
      {start:130000001, end:140000000, value:1},
      {start:140000001, end:150000000, value:2},
      {start:150000001, end:160000000, value:3},
      {start:160000001, end:170000000, value:4},
      {start:170000001, end:180000000, value:5},
      {start:180000001, end:190000000, value:6},
      {start:190000001, end:198022430, value:6}
    ]
  },
  {
    parent: "chr4",
    data: [
      {start:1, end:10000000, value:7},
      {start:10000001, end:20000000, value:8},
      {start:20000001, end:30000000, value:9},
      {start:30000001, end:40000000, value:9},
      {start:40000001, end:50000000, value:8},
      {start:50000001, end:60000000, value:7},
      {start:60000001, end:70000000, value:6},
      {start:70000001, end:80000000, value:5},
      {start:80000001, end:90000000, value:4},
      {start:90000001, end:100000000, value:3},
      {start:100000001, end:110000000, value:2},
      {start:110000001, end:120000000, value:1},
      {start:120000001, end:130000000, value:0},
      {start:130000001, end:140000000, value:0},
      {start:140000001, end:150000000, value:1},
      {start:150000001, end:160000000, value:2},
      {start:160000001, end:170000000, value:3},
      {start:170000001, end:180000000, value:4},
      {start:180000001, end:190000000, value:5},
      {start:190000001, end:191154276, value:5}
    ]
  },
  {
    parent: "chr5",
    data: [
      {start:1, end:10000000, value:6},
      {start:10000001, end:20000000, value:7},
      {start:20000001, end:30000000, value:8},
      {start:30000001, end:40000000, value:9},
      {start:40000001, end:50000000, value:9},
      {start:50000001, end:60000000, value:8},
      {start:60000001, end:70000000, value:7},
      {start:70000001, end:80000000, value:6},
      {start:80000001, end:90000000, value:5},
      {start:90000001, end:100000000, value:4},
      {start:100000001, end:110000000, value:3},
      {start:110000001, end:120000000, value:2},
      {start:120000001, end:130000000, value:1},
      {start:130000001, end:140000000, value:0},
      {start:140000001, end:150000000, value:0},
      {start:150000001, end:160000000, value:1},
      {start:160000001, end:170000000, value:2},
      {start:170000001, end:180000000, value:3},
      {start:180000001, end:180915260, value:3}
    ]
  },
  {
    parent: "chr6",
    data: [
      {start:1, end:10000000, value:4},
      {start:10000001, end:20000000, value:5},
      {start:20000001, end:30000000, value:6},
      {start:30000001, end:40000000, value:7},
      {start:40000001, end:50000000, value:8},
      {start:50000001, end:60000000, value:9},
      {start:60000001, end:70000000, value:9},
      {start:70000001, end:80000000, value:8},
      {start:80000001, end:90000000, value:7},
      {start:90000001, end:100000000, value:6},
      {start:100000001, end:110000000, value:5},
      {start:110000001, end:120000000, value:4},
      {start:120000001, end:130000000, value:3},
      {start:130000001, end:140000000, value:2},
      {start:140000001, end:150000000, value:1},
      {start:150000001, end:160000000, value:0},
      {start:160000001, end:170000000, value:0},
      {start:170000001, end:171115067, value:0}
    ]
  },
  {
    parent: "chr7",
    data: [
      {start:1, end:10000000, value:1},
      {start:10000001, end:20000000, value:2},
      {start:20000001, end:30000000, value:3},
      {start:30000001, end:40000000, value:4},
      {start:40000001, end:50000000, value:5},
      {start:50000001, end:60000000, value:6},
      {start:60000001, end:70000000, value:7},
      {start:70000001, end:80000000, value:8},
      {start:80000001, end:90000000, value:9},
      {start:90000001, end:100000000, value:9},
      {start:100000001, end:110000000, value:8},
      {start:110000001, end:120000000, value:7},
      {start:120000001, end:130000000, value:6},
      {start:130000001, end:140000000, value:5},
      {start:140000001, end:150000000, value:4},
      {start:150000001, end:159138663, value:4}
    ]
  },
  {
    parent: "chr8",
    data: [
      {start:1, end:10000000, value:3},
      {start:10000001, end:20000000, value:2},
      {start:20000001, end:30000000, value:1},
      {start:30000001, end:40000000, value:0},
      {start:40000001, end:50000000, value:0},
      {start:50000001, end:60000000, value:1},
      {start:60000001, end:70000000, value:2},
      {start:70000001, end:80000000, value:3},
      {start:80000001, end:90000000, value:4},
      {start:90000001, end:100000000, value:5},
      {start:100000001, end:110000000, value:6},
      {start:110000001, end:120000000, value:7},
      {start:120000001, end:130000000, value:8},
      {start:130000001, end:140000000, value:9},
      {start:140000001, end:146364022, value:9}
    ]
  },
  {
    parent: "chr9",
    data: [
      {start:1, end:10000000, value:9},
      {start:10000001, end:20000000, value:8},
      {start:20000001, end:30000000, value:7},
      {start:30000001, end:40000000, value:6},
      {start:40000001, end:50000000, value:5},
      {start:50000001, end:60000000, value:4},
      {start:60000001, end:70000000, value:3},
      {start:70000001, end:80000000, value:2},
      {start:80000001, end:90000000, value:1},
      {start:90000001, end:100000000, value:0},
      {start:100000001, end:110000000, value:0},
      {start:110000001, end:120000000, value:1},
      {start:120000001, end:130000000, value:2},
      {start:130000001, end:140000000, value:3},
      {start:140000001, end:141213431, value:3}
    ]
  },
  {
    parent: "chr10",
    data: [
      {start:1, end:10000000, value:4},
      {start:10000001, end:20000000, value:5},
      {start:20000001, end:30000000, value:6},
      {start:30000001, end:40000000, value:7},
      {start:40000001, end:50000000, value:8},
      {start:50000001, end:60000000, value:9},
      {start:60000001, end:70000000, value:9},
      {start:70000001, end:80000000, value:8},
      {start:80000001, end:90000000, value:7},
      {start:90000001, end:100000000, value:6},
      {start:100000001, end:110000000, value:5},
      {start:110000001, end:120000000, value:4},
      {start:120000001, end:130000000, value:3},
      {start:130000001, end:135534747, value:3}
    ]
  },
  {
    parent: "chr11",
    data: [
      {start:1, end:10000000, value:2},
      {start:10000001, end:20000000, value:1},
      {start:20000001, end:30000000, value:0},
      {start:30000001, end:40000000, value:0},
      {start:40000001, end:50000000, value:1},
      {start:50000001, end:60000000, value:2},
      {start:60000001, end:70000000, value:3},
      {start:70000001, end:80000000, value:4},
      {start:80000001, end:90000000, value:5},
      {start:90000001, end:100000000, value:6},
      {start:100000001, end:110000000, value:7},
      {start:110000001, end:120000000, value:8},
      {start:120000001, end:130000000, value:9},
      {start:130000001, end:135006516, value:9}
    ]
  },
  {
    parent: "chr12",
    data: [
      {start:1, end:10000000, value:9},
      {start:10000001, end:20000000, value:8},
      {start:20000001, end:30000000, value:7},
      {start:30000001, end:40000000, value:6},
      {start:40000001, end:50000000, value:5},
      {start:50000001, end:60000000, value:4},
      {start:60000001, end:70000000, value:3},
      {start:70000001, end:80000000, value:2},
      {start:80000001, end:90000000, value:1},
      {start:90000001, end:100000000, value:0},
      {start:100000001, end:110000000, value:0},
      {start:110000001, end:120000000, value:1},
      {start:120000001, end:130000000, value:2},
      {start:130000001, end:133851895, value:2}
    ]
  },
  {
    parent: "chr13",
    data: [
      {start:1, end:10000000, value:3},
      {start:10000001, end:20000000, value:4},
      {start:20000001, end:30000000, value:5},
      {start:30000001, end:40000000, value:6},
      {start:40000001, end:50000000, value:7},
      {start:50000001, end:60000000, value:8},
      {start:60000001, end:70000000, value:9},
      {start:70000001, end:80000000, value:9},
      {start:80000001, end:90000000, value:8},
      {start:90000001, end:100000000, value:7},
      {start:100000001, end:110000000, value:6},
      {start:110000001, end:115169878, value:6}
    ]
  },
  {
    parent: "chr14",
    data: [
      {start:1, end:10000000, value:5},
      {start:10000001, end:20000000, value:4},
      {start:20000001, end:30000000, value:3},
      {start:30000001, end:40000000, value:2},
      {start:40000001, end:50000000, value:1},
      {start:50000001, end:60000000, value:0},
      {start:60000001, end:70000000, value:0},
      {start:70000001, end:80000000, value:1},
      {start:80000001, end:90000000, value:2},
      {start:90000001, end:100000000, value:3},
      {start:100000001, end:107349540, value:3}
    ]
  },
  {
    parent: "chr15",
    data: [
      {start:1, end:10000000, value:4},
      {start:10000001, end:20000000, value:5},
      {start:20000001, end:30000000, value:6},
      {start:30000001, end:40000000, value:7},
      {start:40000001, end:50000000, value:8},
      {start:50000001, end:60000000, value:9},
      {start:60000001, end:70000000, value:9},
      {start:70000001, end:80000000, value:8},
      {start:80000001, end:90000000, value:7},
      {start:90000001, end:100000000, value:6},
      {start:100000001, end:102531392, value:6}
    ]
  },
  {
    parent: "chr16",
    data: [
      {start:1, end:10000000, value:5},
      {start:10000001, end:20000000, value:4},
      {start:20000001, end:30000000, value:3},
      {start:30000001, end:40000000, value:2},
      {start:40000001, end:50000000, value:1},
      {start:50000001, end:60000000, value:0},
      {start:60000001, end:70000000, value:0},
      {start:70000001, end:80000000, value:1},
      {start:80000001, end:90000000, value:2},
      {start:90000001, end:90354753, value:2}
    ]
  },
  {
    parent: "chr17",
    data: [
      {start:1, end:10000000, value:3},
      {start:10000001, end:20000000, value:4},
      {start:20000001, end:30000000, value:5},
      {start:30000001, end:40000000, value:6},
      {start:40000001, end:50000000, value:7},
      {start:50000001, end:60000000, value:8},
      {start:60000001, end:70000000, value:9},
      {start:70000001, end:80000000, value:9},
      {start:80000001, end:81195210, value:9}
    ]
  },
  {
    parent: "chr18",
    data: [
      {start:1, end:10000000, value:8},
      {start:10000001, end:20000000, value:7},
      {start:20000001, end:30000000, value:6},
      {start:30000001, end:40000000, value:5},
      {start:40000001, end:50000000, value:4},
      {start:50000001, end:60000000, value:3},
      {start:60000001, end:70000000, value:2},
      {start:70000001, end:78077248, value:2}
    ]
  },
  {
    parent: "chr19",
    data: [
      {start:1, end:10000000, value:1},
      {start:10000001, end:20000000, value:0},
      {start:20000001, end:30000000, value:0},
      {start:30000001, end:40000000, value:1},
      {start:40000001, end:50000000, value:2},
      {start:50000001, end:59128983, value:2}
    ]
  },
  {
    parent: "chr20",
    data: [
      {start:1, end:10000000, value:3},
      {start:10000001, end:20000000, value:4},
      {start:20000001, end:30000000, value:5},
      {start:30000001, end:40000000, value:6},
      {start:40000001, end:50000000, value:7},
      {start:50000001, end:60000000, value:8},
      {start:60000001, end:63025520, value:8}
    ]
  },
  {
    parent: "chr21",
    data: [
      {start:1, end:10000000, value:9},
      {start:10000001, end:20000000, value:9},
      {start:20000001, end:30000000, value:8},
      {start:30000001, end:40000000, value:7},
      {start:40000001, end:48129895, value:7}
    ]
  },
  {
    parent: "chr22",
    data: [
      {start:1, end:10000000, value:6},
      {start:10000001, end:20000000, value:5},
      {start:20000001, end:30000000, value:4},
      {start:30000001, end:40000000, value:3},
      {start:40000001, end:50000000, value:2},
      {start:50000001, end:51304566, value:2}
    ]
  },
  {
    parent: "chrX",
    data: [
      {start:1, end:10000000, value:1},
      {start:10000001, end:20000000, value:0},
      {start:20000001, end:30000000, value:0},
      {start:30000001, end:40000000, value:1},
      {start:40000001, end:50000000, value:2},
      {start:50000001, end:60000000, value:3},
      {start:60000001, end:70000000, value:4},
      {start:70000001, end:80000000, value:5},
      {start:80000001, end:90000000, value:6},
      {start:90000001, end:100000000, value:7},
      {start:100000001, end:110000000, value:8},
      {start:110000001, end:120000000, value:9},
      {start:120000001, end:130000000, value:9},
      {start:130000001, end:140000000, value:8},
      {start:140000001, end:150000000, value:7},
      {start:150000001, end:155270560, value:7}
    ]
  },
  {
    parent: "chrY",
    data: [
      {start:1, end:10000000, value:6},
      {start:10000001, end:20000000, value:5},
      {start:20000001, end:30000000, value:4},
      {start:30000001, end:40000000, value:3},
      {start:40000001, end:50000000, value:2},
      {start:50000001, end:59373566, value:2}
    ]
  },
]

links = [
  ['january', 1, 10, 'february', 20, 25, 1],
  ['february', 1, 10, 'march', 20, 25, 1],
  ['march', 1, 10, 'april', 3, 7, 2],
  ['april', 1, 10, 'may', 20, 25, 3],
  ['may', 1, 10, 'june', 20, 25, 4],
  ['june', 1, 10, 'july', 20, 25, 5],
  ['july', 1, 10, 'august', 20, 25, 6],
  ['august', 1, 10, 'september', 20, 25, 7],
  ['september', 1, 10, 'october', 20, 25, 8],
  ['october', 1, 10, 'november', 20, 25, 9],
  ['november', 1, 10, 'december', 20, 25, 10],
  ['december', 1, 10, 'january', 20, 25, 11],
];
linkRules = [
  {
    parameter: 'color',
    value: 'red',
    condition: function(datum, i){
      return datum.source.id == 'december'
    }
  },
  {
    parameter: 'opacity',
    value: 0.2,
    condition: function(datum, i){
      return datum.target.id == 'february'
    }
  }
];

scatter = [
  ['january', 1, 6],
  ['january', 2, 2],
  ['january', 3, 4],
  ['january', 4, 1],
  ['january', 5, 3],
  ['january', 6, 6],
  ['january', 7, 5],
  ['january', 8, 4],
  ['january', 9, 6],
  ['january', 31, 13],
  ['february', 1, 6],
  ['february', 2, 3],
  ['february', 3, 4],
  ['february', 4, 2],
  ['february', 5, 6],
  ['february', 6, 5]
]

stack = [
  ['january', 1, 10, 8],
  ['january', 1, 20, 7],
  ['january', 11, 23, 3],
  ['january', 2, 20, 4],
  ['january', 19, 25, 5],
  ['january', 21, 25, 6]
]

heatmap = [
  // ['january', 'a', 2,3],
  ['january', 1, 4,2],
  ['january', 5, 10,10],
  ['january', 11, 15,13],
  ['february', 1, 28,3],
  ['march', 1, 2,3],
  ['march', 1, 2,3]
];

highlight = [
  ['january', 1, 2,'blue'],
  ['january', 1, 4,'red'],
  ['january', 5, 10,'green'],
  ['january', 11, 15,'yellow'],
  ['february', 1, 28,'grey'],
  ['march', 1, 2,'purple'],
  ['march', 1, 2,'black']
];

var circos = new Circos({
  container: '#chart'
});

var backgrounds = [
  {start: 0, end: 0.5, color: '#d3d3d3', opacity: 1},
  {start: 0.5, end: 1, color: '#333333', opacity: 0.5, parent: ['february', 'april']}
]

var stack_rules = [
  // {parameter: 'color', value: 'red', condition: function(parent, datum, layer_id){ return (datum.end - datum.start > 8);}},
  // {parameter: 'color', value: 'blue', condition: function(parent, datum, layer_id){ return (layer_id > 2);}}
  // {parameter: 'usePalette', value: true, condition: function(parent, datum, layer_id){ return (layer_id > 2);}}
]

backgrounds2 = [
  {start: 0, end: 0.5, color: '#d3d3d3'},
  {start: 0.5, end: 0.8, color: '#fdbb84', opacity: 0.8},
  {start: 0.8, end: 1, color: '#e34a33', parent: ['january', 'february']},
  {start: 0.8, end: 1, color: '#2c7fb8', parent: 'march'}
];

text = [
  ['january', 10, 'toto'],
  ['february', 3, 'titi'],
  ['may', 25, 'tuti']
]

circos
  .layout(
    {
      ticks: {display: false},
      cornerRadius: 0,
      onClick: function(){console.log('You clicked on the layout layer!');}
    },
    layout_data
  )
  // .text('text1', {innerRadius: 180, style: function(d, i){
  //   var color = i%2 === 0 ? 'blue' : 'green';
  //   return {'font-size': '50px', 'font-family': 'Arial', 'color': color};
  // }}, text)
  // .heatmap('h1', {
  //   usePalette: false,
  //   color: function(d, i){
  //     if (i%2 == 0) {
  //       return 'green';
  //     }
  //   },
  //   tooltipContent: function(d){
  //     return 'From ' + d.start + ' to ' + d.end + ' ' + d.block_id + ': ' + d.value;
  //   }
  // }, heatmap)
  // .histogram(
  //   'hist1',
  //   {
  //     innerRadius: 200,
  //     outerRadius: 220,
  //     axes: {display: true},
  //     usePalette: false,
  //     color: function(d, i) {
  //       return i % 2 === 0 ? 'blue' : 'red';
  //     },
  //   }, heatmap)
  // .scatter(
  //   'c1',
  //   {
  //     innerRadius: 150,
  //     outerRadius: 200,
  //     glyph: {
  //       shape: function(d, i) {
  //         if (d.block_id === 'january') {
  //           return 'cross'
  //         }
  //         return 'circle'
  //       },
  //       size: 50,
  //       fill: true,
  //       color: function(d, i) {
  //         return i % 2 === 0 ? 'blue' : 'red'
  //       }
  //     }
  //   },
  //   scatter
  // )
  // .line(
  //   'line1',
  //   {
  //     innerRadius: 150,
  //     outerRadius: 200,
  //     direction: 'in',
  //     fill: true,
  //     fill_color: function(blockId, i) {
  //       return blockId === 'january' ? 'blue' : 'red';
  //     }
  //   },
  //   scatter
  // )
  .chords(
    'l1',
    {
      logScale: false,
      opacity: 0.7,
      color: 'Greens',
      tooltipContent: function(d) {
        return d.source.id + ' ➤ ' + d.target.id + ': ' + d.value;
      },
    },
    links
  )
  // .stack('stack1', {thickness: 10, usePalette: true, margin: 0, direction: 'center', innerRadius: 150, outerRadius: 240, axes: {display: true}}, stack)
  // .stack('stack1', {thickness: 10, usePalette: true, margin: 0, direction: 'out', innerRadius: 150, outerRadius: 240}, stack)
  // .line('line2', {innerRadius:100, outerRadius: 200}, [], {}, backgrounds2)
  // .highlight('highlight1', {
  //   innerRadius: 100, outerRadius: 200, opacity: 1
  // }, highlight)
  .render();

// circos.removeTracks(['l1'])
// circos.removeTracks('stack1')
// circos.removeTracks()
  // .render(['layout', 'h1']);
